File: 07_app_development/MyProject/libtsparser/src/main/java/com/tosmart/xiexuming/libtsparser/TransportStreamManager.java
package com.tosmart.xiexuming.libtsparser;

import com.tosmart.xiexuming.libtsparser.program.Program;
import com.tosmart.xiexuming.libtsparser.program.ProgramManager;
import com.tosmart.xiexuming.libtsparser.section.ProgramAssociationSection;
import com.tosmart.xiexuming.libtsparser.section.ProgramAssociationSectionManager;
import com.tosmart.xiexuming.libtsparser.section.ProgramMapSection;
import com.tosmart.xiexuming.libtsparser.section.ProgramMapSectionManager;
import com.tosmart.xiexuming.libtsparser.section.SectionListener;
import com.tosmart.xiexuming.libtsparser.section.ServiceDescriptionSection;
import com.tosmart.xiexuming.libtsparser.section.ServiceDescriptionSectionManager;
import com.tosmart.xiexuming.libtsparser.table.ProgramAssociationTable;
import com.tosmart.xiexuming.libtsparser.table.ProgramAssociationTableManager;
import com.tosmart.xiexuming.libtsparser.table.ProgramMapTable;
import com.tosmart.xiexuming.libtsparser.table.ProgramMapTableManager;
import com.tosmart.xiexuming.libtsparser.table.ServiceDescriptionTable;
import com.tosmart.xiexuming.libtsparser.table.ServiceDescriptionTableManager;
import com.tosmart.xiexuming.libtsparser.tspacket.TsPacketManager;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;


/**
 * @author xiexuming
 */
public class TransportStreamManager {
    ProgramAssociationSectionManager mProgramAssociationSectionManager;
    ProgramMapSectionManager mProgramMapSectionManager;
    ServiceDescriptionSectionManager mServiceDescriptionSectionManager;

    ProgramAssociationTableManager mProgramAssociationTableManager;
    ProgramMapTableManager mProgramMapTableManager;
    ServiceDescriptionTableManager mServiceDescriptionTableManager;

    HashMap<Integer, ProgramMapTable> mProgramMapTableMap;
    ServiceDescriptionTable mServiceDescriptionTable;
    ProgramAssociationTable mProgramAssociationTable;
    TsPacketManager mPacketManager;

    public TransportStreamManager() {
        mPacketManager = TsPacketManager.getInstance();
        mProgramMapTableMap = new HashMap<>();
        mServiceDescriptionTableManager = new ServiceDescriptionTableManager();
        mProgramAssociationSectionManager = new ProgramAssociationSectionManager();
        mProgramMapSectionManager = new ProgramMapSectionManager();
        mServiceDescriptionSectionManager = new ServiceDescriptionSectionManager();
        mProgramAssociationTableManager = new ProgramAssociationTableManager();
        mProgramMapTableManager = new ProgramMapTableManager();
    }

    public List<Program> parseTsStream(String filePath) {
        try {
            int tsPacketLength = mPacketManager.getTsPacketLength(filePath);
            System.out.println(tsPacketLength);
            if (tsPacketLength <= 0) {
                return null;
            }
            SectionListener sectionListener = new SectionListener() {
                @Override
                public void makePat(ProgramAssociationSection programAssociationSection) {
                    boolean programAssociationComposeFinish = mProgramAssociationTableManager.composeSection(programAssociationSection);
                    if (programAssociationComposeFinish) {
                        mProgramAssociationTable =
                                mProgramAssociationTableManager.getProgramAssociationTable();
                        ArrayList<Integer> programPidList = mProgramAssociationTableManager.getProgramPidList();
                        mProgramMapSectionManager.setFilterPidList(programPidList);
                        mPacketManager.deleteObserver(mProgramAssociationSectionManager);
                    }
                }

                @Override
                public void makePmt(int pid, ProgramMapSection programMapSection) {
                    ProgramMapTable programMapTable = mProgramMapTableManager.getProgramMapTable(programMapSection, pid);
                    mProgramMapTableMap.put(pid, programMapTable);
                    ArrayList<Integer> filterPidList = mProgramMapSectionManager.getFilterPidList();
                    if (filterPidList.size() <= 0) {
                        mPacketManager.deleteObserver(mProgramMapSectionManager);
                    }
                }

                @Override
                public void makeSdt(ServiceDescriptionSection serviceDescriptionSection) {
                    boolean serviceDescriptionComposeFinish = mServiceDescriptionTableManager.composeSection(serviceDescriptionSection);
                    if (serviceDescriptionComposeFinish) {
                        mServiceDescriptionTable = mServiceDescriptionTableManager.getServiceDescriptionTable();
                        mPacketManager.deleteObserver(mServiceDescriptionSectionManager);
                    }
                }
            };

            mProgramAssociationSectionManager.setListener(sectionListener);
            mProgramMapSectionManager.setListener(sectionListener);
            mServiceDescriptionSectionManager.setListener(sectionListener);

            mPacketManager.addObserver(mProgramMapSectionManager);
            mPacketManager.addObserver(mProgramAssociationSectionManager);
            mPacketManager.addObserver(mServiceDescriptionSectionManager);
            mPacketManager.distributePacket(filePath);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return ProgramManager.getInstance().getProgramList(mProgramAssociationTable, mServiceDescriptionTable, mProgramMapTableMap);
    }

    public void forceClose() {
        mPacketManager.deleteObservers();
    }
}

File: newpaserapp/libtosmartparse/src/main/java/com/example/libtosmartparse/TransportStreamManager.kt
import com.example.libtosmartparse.packet.TsPacketManager
import com.example.libtosmartparse.section.*
import com.example.libtosmartparse.table.*
import java.util.*

/**
 * @author xxm
 * 2022/7/18
 **/
class TransportStreamManager {
    var mProgramAssociationSectionManager: ProgramAssociationSectionManager
    var mProgramMapSectionManager: ProgramMapSectionManager
    var mServiceDescriptionSectionManager: ServiceDescriptionSectionManager
    var mProgramAssociationTableManager: ProgramAssociationTableManager
    var mProgramMapTableManager: ProgramMapTableManager
    var mServiceDescriptionTableManager: ServiceDescriptionTableManager
    var mProgramMapTableMap: HashMap<Int, ProgramMapTable>? = null
    var mServiceDescriptionTable: ServiceDescriptionTable? = null
    var mProgramAssociationTable: ProgramAssociationTable? = null
    var mPacketManager: TsPacketManager

    fun parseTsStream(filePath: String?): List<Program>? {
        try {
            val tsPacketLength: Int =
                if (filePath.isNullOrBlank()) 0 else mPacketManager.getTsPacketLength(filePath)
            if (tsPacketLength <= 0) {
                return null
            }
            val sectionListener: SectionListener = object : SectionListener {
                override fun makePat(programAssociationSection: ProgramAssociationSection?) {
                    val programAssociationComposeFinish: Boolean =
                        mProgramAssociationTableManager.composeSection(programAssociationSection)
                    if (programAssociationComposeFinish) {
                        mProgramAssociationTable =
                            mProgramAssociationTableManager.getProgramAssociationTable()
                        val programPidList: ArrayList<Int> =
                            mProgramAssociationTableManager.getProgramPidList()
                        mProgramMapSectionManager.setFilterPidList(programPidList)
                        mPacketManager.deleteObserver(mProgramAssociationSectionManager)
                    }
                }

                override fun makePmt(pid: Int, programMapSection: ProgramMapSection?) {
                    val programMapTable: ProgramMapTable =
                        mProgramMapTableManager.getProgramMapTable(programMapSection!!, pid)
                    mProgramMapTableMap?.set(pid, programMapTable)
                    val filterPidList: ArrayList<Int> = mProgramMapSectionManager.getFilterPidList()
                    if (filterPidList.size <= 0) {
                        mPacketManager.deleteObserver(mProgramMapSectionManager)
                    }
                }

                override fun makeSdt(serviceDescriptionSection: ServiceDescriptionSection?) {
                    val serviceDescriptionComposeFinish: Boolean =
                        mServiceDescriptionTableManager.composeSection(serviceDescriptionSection)
                    if (serviceDescriptionComposeFinish) {
                        mServiceDescriptionTable =
                            mServiceDescriptionTableManager.getServiceDescriptionTable()
                        mPacketManager.deleteObserver(mServiceDescriptionSectionManager)
                    }
                }
            }

            mProgramAssociationSectionManager.setListener(sectionListener)
            mProgramMapSectionManager.setListener(sectionListener)
            mServiceDescriptionSectionManager.setListener(sectionListener)
            mPacketManager.addObserver(mProgramMapSectionManager)
            mPacketManager.addObserver(mProgramAssociationSectionManager)
            mPacketManager.addObserver(mServiceDescriptionSectionManager)
            mPacketManager.distributePacket(filePath)
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return ProgramManager.instance
            ?.getProgramList(
                mProgramAssociationTable,
                mServiceDescriptionTable, mProgramMapTableMap
            )
    }

    fun forceClose() {
        mPacketManager.deleteObservers()
    }

    init {
        mPacketManager = TsPacketManager.getInstance() as TsPacketManager
        mProgramMapTableMap = HashMap<Int, ProgramMapTable>()
        mServiceDescriptionTableManager = ServiceDescriptionTableManager()
        mProgramAssociationSectionManager = ProgramAssociationSectionManager()
        mProgramMapSectionManager = ProgramMapSectionManager()
        mServiceDescriptionSectionManager = ServiceDescriptionSectionManager()
        mProgramAssociationTableManager = ProgramAssociationTableManager()
        mProgramMapTableManager = ProgramMapTableManager()
    }
}
